crates/sui-core/src/epoch/randomness.rs

```diff
@@ -0,0 +1,512 @@
+use std::collections::BTreeMap;
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use std::sync::{Arc, Mutex, Weak};
+
+use fastcrypto::encoding::{Encoding, Hex};
+use fastcrypto::groups::bls12381;
+use fastcrypto::serde_helpers::ToFromByteArray;
+use fastcrypto::traits::{KeyPair, ToFromBytes};
+use fastcrypto_tbls::nodes::PartyId;
+use fastcrypto_tbls::{dkg, nodes};
+use sui_types::base_types::AuthorityName;
+use sui_types::committee::{Committee, StakeUnit};
+use sui_types::crypto::AuthorityKeyPair;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages_consensus::ConsensusTransaction;
+use tokio::sync::OnceCell;
+use tracing::{debug, error, info};
+use typed_store::rocks::DBBatch;
+use typed_store::Map;
+
+use crate::authority::authority_per_epoch_store::{AuthorityEpochTables, AuthorityPerEpochStore};
+use crate::consensus_adapter::ConsensusAdapter;
+
+type PkG = bls12381::G2Element;
+type EncG = bls12381::G2Element;
+
+const SINGLETON_KEY: u64 = 0;
+
+// State machine for randomness DKG and generation.
+//
+// DKG protocol:
+// 1. This node sends out a `Message` to all other nodes.
+// 2. Once sufficient valid `Message`s are received from other nodes via consensus and processed,
+//    this node sends out a `Confirmation` to all other nodes.
+// 3. Once sufficient `Confirmation`s are received from other nodes via consensus and processed,
+//    they are combined to form a public VSS key and local private key shares.
+// 4. Randomness generation begins.
+//
+// Randomness generation:
+// TODO: document when implemented.
+pub struct RandomnessManager {
+    epoch_store: Weak<AuthorityPerEpochStore>,
```

Comment:
I may be wrong, but it looks to me like everywhere `epoch_store` is used, it could have been passed in as a function arg

Reply:
Yeah, I suppose so. Doing it this way seemed maybe slightly cleaner than having `epoch_store` arguments on every public fn here, but if you feel strongly I could switch it to the latter. I could see a case for either.